horovod/run/common/util/safe_shell_exec.py

```diff
@@ -69,75 +74,104 @@ def prepend_context(line, rank, prefix):
             line=line
         )
 
-    with os.fdopen(src_fd, 'r') as src:
-        line_buffer = ''
-        while True:
-            text = os.read(src.fileno(), 1000)
-            if not isinstance(text, str):
-                text = text.decode('utf-8')
-            if not text:
-                break
+    line_buffer = ''
+    while True:
+        text = os.read(src_stream.fileno(), 1000)
+        if not isinstance(text, str):
+            text = text.decode('utf-8')
+        if not text or len(text) == 0:
+            break
 
-            for line in re.split('([\r\n])', text):
-                line_buffer += line
-                if line == '\r' or line == '\n':
-                    if index is not None:
-                        line_buffer = prepend_context(line_buffer, index, prefix)
+        for line in re.split('([\r\n])', text):
+            line_buffer += line
+            if line == '\r' or line == '\n':
+                if index is not None:
+                    line_buffer = prepend_context(line_buffer, index, prefix)
 
-                    dst_stream.write(line_buffer)
-                    dst_stream.flush()
-                    line_buffer = ''
+                dst_stream.write(line_buffer)
+                dst_stream.flush()
+                line_buffer = ''
 
-        # flush the line buffer if it is not empty
-        if len(line_buffer):
-            if index is not None:
-                line_buffer = prepend_context(line_buffer, index, prefix)
-            dst_stream.write(line_buffer)
-            dst_stream.flush()
+    # flush the line buffer if it is not empty
+    if len(line_buffer):
+        if index is not None:
```

Comment:
these four lines in this `if` body are identical to the for loop, can them be moved into a `def`?